x/wasm/internal/keeper/keeper_test.go

```diff
@@ -451,6 +453,116 @@ func TestExecuteWithStorageLoop(t *testing.T) {
+	fred := createFakeFundedAccount(ctx, accKeeper, topUp)
+
+	wasmCode, err := ioutil.ReadFile("./testdata/contract.wasm")
+	require.NoError(t, err)
+
+	originalContractID, err := keeper.Create(ctx, creator, wasmCode, "", "")
+	require.NoError(t, err)
+	newContractID, err := keeper.Create(ctx, creator, wasmCode, "", "")
+	require.NoError(t, err)
+
+	_, _, anyAddr := keyPubAddr()
+	initMsg := InitMsg{
+		Verifier:    fred,
+		Beneficiary: anyAddr,
+	}
+	initMsgBz, err := json.Marshal(initMsg)
+	require.NoError(t, err)
+	specs := map[string]struct {
+		admin                sdk.AccAddress
+		overrideContractAddr sdk.AccAddress
+		caller               sdk.AccAddress
+		codeID               uint64
+		migrateMsg           []byte
+		expErr               *sdkerrors.Error
+	}{
+		"all good with same code id": {
+			admin:  creator,
+			caller: creator,
+			codeID: originalContractID,
+		},
+		"all good with new code id": {
+			admin:  creator,
+			caller: creator,
+			codeID: newContractID,
+		},
+		"prevent migration when admin was not set on instantiate": {
+			caller: creator,
+			codeID: originalContractID,
+			expErr: sdkerrors.ErrUnauthorized,
+		},
+		"prevent migration when not admin": {
+			caller: creator,
+			admin:  fred,
+			codeID: originalContractID,
+			expErr: sdkerrors.ErrUnauthorized,
+		},
+		"fail with non existing code id": {
+			admin:  creator,
+			caller: creator,
+			codeID: 99999,
+			expErr: sdkerrors.ErrInvalidRequest,
+		},
+		"fail with non existing contract addr": {
+			admin:                creator,
+			caller:               creator,
+			overrideContractAddr: anyAddr,
+			codeID:               originalContractID,
+			expErr:               sdkerrors.ErrInvalidRequest,
+		},
+		"fail when migration caused error": {
+			admin:                creator,
+			caller:               creator,
+			overrideContractAddr: anyAddr,
+			codeID:               originalContractID,
+			migrateMsg:           bytes.Repeat([]byte{0x1}, 7), // condition hard coded in stub: >6 = error
+			expErr:               sdkerrors.ErrInvalidRequest,
+		},
+	}
+	var (
+		builtIntoGoCosmWasmStubGas  = sdk.Gas(10000)
+		builtIntoGoCosmWasmStubData = []byte(("my-migration-response-data"))
+	)
+	for msg, spec := range specs {
+		t.Run(msg, func(t *testing.T) {
+			ctx = ctx.WithBlockHeight(ctx.BlockHeight() + 1)
+			addr, err := keeper.Instantiate(ctx, originalContractID, creator, spec.admin, initMsgBz, "demo contract", nil)
+			require.NoError(t, err)
+			if spec.overrideContractAddr != nil {
+				addr = spec.overrideContractAddr
+			}
+			gasBefore := ctx.GasMeter().GasConsumed()
+			res, err := keeper.Migrate(ctx, addr, spec.caller, spec.codeID, spec.migrateMsg)
+			require.True(t, spec.expErr.Is(err), "expected %v but got %+v", spec.expErr, err)
+			if spec.expErr != nil {
+				return
+			}
+			gasAfter := ctx.GasMeter().GasConsumed()
+			assert.Greater(t, gasAfter-gasBefore, builtIntoGoCosmWasmStubGas)
```

Comment:
this is a bit misleading....

the wasm contract returns "consumed wasm gas", which is not the same as the sdk gas (measured by the gas meter).

You can see here https://github.com/CosmWasm/wasmd/blob/0.9/x/wasm/internal/keeper/keeper.go#L341-L353 how we use a multiplicatin factor to translate. Currently 100 wasm gas = 1 sdk gas, but I do think that is too cheap.

In any case, the execution uses quite a bit of sdk gas accessing the storage to read and write contract and code info, so it is over 10k. If the wasm contract returned 1 million, that would be 10k sdk gas + the storage usage < 1 million and this check would break.

You can divide `builtIntoGoCosmWasmStubData` by `GasMultiplier` and this is guaranteed to be valid.